File: crates/krishiv-python/examples/batch_sql.py
#!/usr/bin/env python3
"""Standard batch SQL dataframe query example in Python."""

import os
import tempfile
import pyarrow as pa
import pyarrow.parquet as pq
import krishiv as ks

def main():
    with tempfile.TemporaryDirectory() as temp_dir:
        parquet_path = os.path.join(temp_dir, "people.parquet")
        write_people_parquet(parquet_path)

        session = ks.Session.from_env()
        session.register_parquet("people", parquet_path)

        result = (
            session.sql("select city, count(*) as count from people group by city order by city")
            .collect()
        )
        print(result.pretty())

def write_people_parquet(path):
    schema = pa.schema([
        ("id", pa.int64()),
        ("city", pa.string()),
    ])
    table = pa.Table.from_pydict({
        "id": [1, 2, 3],
        "city": ["London", "Paris", "London"],
    }, schema=schema)
    pq.write_table(table, path)

if __name__ == "__main__":
    main()

File: crates/krishiv/examples/batch_sql.rs
#![forbid(unsafe_code)]

use std::error::Error;
use std::fs::File;
use std::sync::Arc;

use arrow::array::{Int64Array, StringArray};
use krishiv::{DataType, ExecutionMode, Field, RecordBatch, Schema, Session};
use parquet::arrow::ArrowWriter;
use tempfile::tempdir;

fn main() -> Result<(), Box<dyn Error>> {
    let temp = tempdir()?;
    let parquet_path = temp.path().join("people.parquet");
    write_people_parquet(&parquet_path)?;

    let mut builder = Session::builder();
    if let Ok(url) = std::env::var("KRISHIV_COORDINATOR_URL") {
        builder = builder.with_local_cluster(url);
    } else {
        builder = builder.with_execution_mode(ExecutionMode::Embedded);
    }
    let session = builder.build()?;
    session.register_parquet("people", &parquet_path)?;

    let result = session
        .sql("select city, count(*) as count from people group by city order by city")?
        .collect()?;

    println!("{}", result.pretty()?);
    Ok(())
}

fn write_people_parquet(path: &std::path::Path) -> Result<(), Box<dyn Error>> {
    let schema = Arc::new(Schema::new(vec![
        Field::new("id", DataType::Int64, false),
        Field::new("city", DataType::Utf8, false),
    ]));
    let batch = RecordBatch::try_new(
        schema.clone(),
        vec![
            Arc::new(Int64Array::from(vec![1, 2, 3])),
            Arc::new(StringArray::from(vec!["London", "Paris", "London"])),
        ],
    )?;
    let file = File::create(path)?;
    let mut writer = ArrowWriter::try_new(file, schema, None)?;
    writer.write(&batch)?;
    writer.close()?;
    Ok(())
}
